{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Github\n",
    "\n",
    "The `Github` toolkit contains tools that enable an LLM agent to interact with a github repository. \n",
    "The tool is a wrapper for the [PyGitHub](https://github.com/PyGithub/PyGithub) library. \n",
    "\n",
    "## Quickstart\n",
    "1. Install the pygithub library\n",
    "2. Create a Github app\n",
    "3. Set your environmental variables\n",
    "4. Pass the tools to your agent with `toolkit.get_tools()`\n",
    "\n",
    "Each of these steps will be explained in greate detail below.\n",
    "\n",
    "1. **Get Issues**- fetches issues from the repository.\n",
    "\n",
    "2. **Get Issue**- feteches details about a specific issue.\n",
    "\n",
    "3. **Comment on Issue**- posts a comment on a specific issue.\n",
    "\n",
    "4. **Create Pull Request**- creates a pull request from the bot's working branch to the base branch.\n",
    "\n",
    "5. **Create File**- creates a new file in the repository.\n",
    "\n",
    "6. **Read File**- reads a file from the repository.\n",
    "\n",
    "7. **Update File**- updates a file in the repository.\n",
    "\n",
    "8. **Delete File**- deletes a file from the repository.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Install the `pygithub` library "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%pip install pygithub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Create a Github App\n",
    "\n",
    "[Follow the instructions here](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/registering-a-github-app) to create and register a Github app. Make sure your app has the following [repository permissions:](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28)\n",
    "* Commit statuses (read only)\n",
    "* Contents (read and write)\n",
    "* Issues (read and write)\n",
    "* Metadata (read only)\n",
    "* Pull requests (read and write)\n",
    "\n",
    "\n",
    "\n",
    "Once the app has been registered, add it to the repository you wish the bot to act upon.\n",
    "\n",
    "### 3. Set Environmental Variables\n",
    "\n",
    "Before initializing your agent, the following environmental variables need to be set:\n",
    "\n",
    "* **GITHUB_APP_ID**- A six digit number found in your app's general settings\n",
    "* **GITHUB_APP_PRIVATE_KEY**- The location of your app's private key .pem file\n",
    "* **GITHUB_REPOSITORY**- The name of the Github repository you want your bot to act upon. Must follow the format {username}/{repo-name}. Make sure the app has been added to this repository first!\n",
    "* **GITHUB_BRANCH**- The branch where the bot will make its commits. Defaults to 'master.'\n",
    "* **GITHUB_BASE_BRANCH**- The base branch of your repo, usually either 'main' or 'master.' This is where pull requests will base from. Defaults to 'master.'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example: Simple Agent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from langchain.agents import AgentType\n",
    "from langchain.agents import initialize_agent\n",
    "from langchain.agents.agent_toolkits.github.toolkit import GitHubToolkit\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.utilities.github import GitHubAPIWrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set your environment variables using os.environ\n",
    "os.environ[\"GITHUB_APP_ID\"] = \"123456\"\n",
    "os.environ[\"GITHUB_APP_PRIVATE_KEY\"] = \"path/to/your/private-key.pem\" # or the full .pem string itself\n",
    "os.environ[\"GITHUB_REPOSITORY\"] = \"username/repo-name\"\n",
    "os.environ[\"GITHUB_BRANCH\"] = \"bot-branch-name\"\n",
    "os.environ[\"GITHUB_BASE_BRANCH\"] = \"main\"\n",
    "\n",
    "# This example also requires an OpenAI API key\n",
    "os.environ[\"OPENAI_API_KEY\"] = \"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available tools:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[GitHubAction(name='Get Issues', description=\"\\nThis tool will fetch a list of the repository's issues. It will return the title, and issue number of 5 issues. It takes no input.\", args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.NoInput'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='get_issues'),\n",
       " GitHubAction(name='Get Issue', description='\\nThis tool will fetch the title, body, and comment thread of a specific issue. **VERY IMPORTANT**: You must specify the issue number as an integer.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.GetIssue'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='get_issue'),\n",
       " GitHubAction(name='Comment on Issue', description=\"\\nThis tool is useful when you need to comment on a GitHub issue. Simply pass in the issue number and the comment you would like to make. Please use this sparingly as we don't want to clutter the comment threads. **VERY IMPORTANT**: Your input to this tool MUST strictly follow these rules:\\n\\n- First you must specify the issue number as an integer\\n- Then you must place two newlines\\n- Then you must specify your comment\", args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.CommentOnIssue'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='comment_on_issue'),\n",
       " GitHubAction(name='List open pull requests (PRs)', description=\"\\nThis tool will fetch a list of the repository's Pull Requests (PRs). It will return the title, and PR number of 5 PRs. It takes no input.\", args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.NoInput'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='list_open_pull_requests'),\n",
       " GitHubAction(name='Get Pull Request', description='\\nThis tool will fetch the title, body, comment thread and commit history of a specific Pull Request (by PR number). **VERY IMPORTANT**: You must specify the PR number as an integer.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.GetPR'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='get_pull_request'),\n",
       " GitHubAction(name='Overview of files included in PR', description='\\nThis tool will fetch the full text of all files in a pull request (PR) given the PR number as an input. This is useful for understanding the code changes in a PR or contributing to it. **VERY IMPORTANT**: You must specify the PR number as an integer input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.GetPR'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='list_pull_request_files'),\n",
       " GitHubAction(name='Create Pull Request', description='\\nThis tool is useful when you need to create a new pull request in a GitHub repository. **VERY IMPORTANT**: Your input to this tool MUST strictly follow these rules:\\n\\n- First you must specify the title of the pull request\\n- Then you must place two newlines\\n- Then you must write the body or description of the pull request\\n\\nWhen appropriate, always reference relevant issues in the body by using the syntax `closes #<issue_number` like `closes #3, closes #6`.\\nFor example, if you would like to create a pull request called \"README updates\" with contents \"added contributors\\' names, closes #3\", you would pass in the following string:\\n\\nREADME updates\\n\\nadded contributors\\' names, closes #3', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.CreatePR'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='create_pull_request'),\n",
       " GitHubAction(name=\"List Pull Requests' Files\", description='\\nThis tool will fetch the full text of all files in a pull request (PR) given the PR number as an input. This is useful for understanding the code changes in a PR or contributing to it. **VERY IMPORTANT**: You must specify the PR number as an integer input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.GetPR'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='list_pull_request_files'),\n",
       " GitHubAction(name='Create File', description='\\nThis tool is a wrapper for the GitHub API, useful when you need to create a file in a GitHub repository. **VERY IMPORTANT**: Your input to this tool MUST strictly follow these rules:\\n\\n- First you must specify which file to create by passing a full file path (**IMPORTANT**: the path must not start with a slash)\\n- Then you must specify the contents of the file\\n\\nFor example, if you would like to create a file called /test/test.txt with contents \"test contents\", you would pass in the following string:\\n\\ntest/test.txt\\n\\ntest contents', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.CreateFile'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='create_file'),\n",
       " GitHubAction(name='Read File', description='\\nThis tool is a wrapper for the GitHub API, useful when you need to read the contents of a file. Simply pass in the full file path of the file you would like to read. **IMPORTANT**: the path must not start with a slash', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.ReadFile'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='read_file'),\n",
       " GitHubAction(name='Update File', description='\\nThis tool is a wrapper for the GitHub API, useful when you need to update the contents of a file in a GitHub repository. **VERY IMPORTANT**: Your input to this tool MUST strictly follow these rules:\\n\\n- First you must specify which file to modify by passing a full file path (**IMPORTANT**: the path must not start with a slash)\\n- Then you must specify the old contents which you would like to replace wrapped in OLD <<<< and >>>> OLD\\n- Then you must specify the new contents which you would like to replace the old contents with wrapped in NEW <<<< and >>>> NEW\\n\\nFor example, if you would like to replace the contents of the file /test/test.txt from \"old contents\" to \"new contents\", you would pass in the following string:\\n\\ntest/test.txt\\n\\nThis is text that will not be changed\\nOLD <<<<\\nold contents\\n>>>> OLD\\nNEW <<<<\\nnew contents\\n>>>> NEW', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.UpdateFile'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='update_file'),\n",
       " GitHubAction(name='Delete File', description='\\nThis tool is a wrapper for the GitHub API, useful when you need to delete a file in a GitHub repository. Simply pass in the full file path of the file you would like to delete. **IMPORTANT**: the path must not start with a slash', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.DeleteFile'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='delete_file'),\n",
       " GitHubAction(name='Overview of existing files in Main branch (no input)', description='\\nThis tool will provide an overview of all existing files in the main branch of the repository. It will list the file names, their respective paths, and a brief summary of their contents. This can be useful for understanding the structure and content of the repository, especially when navigating through large codebases. No input parameters are required.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.NoInput'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='list_files_in_main_branch'),\n",
       " GitHubAction(name='Overview of files in your current working branch', description='\\nThis tool will provide an overview of all files in your current working branch where you should implement changes. This is great for getting a high level overview of the structure of your code. No input parameters are required.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.NoInput'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='list_files_in_bot_branch'),\n",
       " GitHubAction(name='List branches in this repository', description='\\nThis tool will fetch a list of all branches in the repository. It will return the name of each branch. No input parameters are required.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.NoInput'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='list_branches_in_repo'),\n",
       " GitHubAction(name='Set active branch', description='\\nThis tool will set the active branch in the repository, similar to `git checkout <branch_name>` and `git switch -c <branch_name>`. **VERY IMPORTANT**: You must specify the name of the branch as a string input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.BranchName'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='set_active_branch'),\n",
       " GitHubAction(name='Create a new branch', description='\\nThis tool will create a new branch in the repository. **VERY IMPORTANT**: You must specify the name of the new branch as a string input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.BranchName'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='create_branch'),\n",
       " GitHubAction(name='Get files from a directory', description='\\nThis tool will fetch a list of all files in a specified directory. **VERY IMPORTANT**: You must specify the path of the directory as a string input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.DirectoryPath'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='get_files_from_directory'),\n",
       " GitHubAction(name='Search issues and pull requests', description='\\nThis tool will search for issues and pull requests in the repository. **VERY IMPORTANT**: You must specify the search query as a string input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.SearchIssuesAndPRs'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='search_issues_and_prs'),\n",
       " GitHubAction(name='Search code', description='\\nThis tool will search for code in the repository. **VERY IMPORTANT**: You must specify the search query as a string input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.SearchCode'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='search_code'),\n",
       " GitHubAction(name='Create review request', description='\\nThis tool will create a review request on the open pull request that matches the current active branch. **VERY IMPORTANT**: You must specify the username of the person who is being requested as a string input parameter.', args_schema=<class 'langchain.agents.agent_toolkits.github.toolkit.CreateReviewRequest'>, api_wrapper=GitHubAPIWrapper(github=<github.MainClass.Github object at 0x12f143490>, github_repo_instance=Repository(full_name=\"KastanDay/ML4Bio\"), github_repository='KastanDay/ML4Bio', github_app_id='373429', github_app_private_key='-----BEGIN RSA PRIVATE KEY-----\\nMIIEpAIBAAKCAQEA7KyK9h2dEe1Ua1w6jgS5Gy4YYzUqyhXmBqoVUe0xidOl1Xsf\\n0uewPSU/83BYFDvt+oSce+r2Ey/LbeuZHU6hhxF5EZRWvwDZMkCpmtiAsr15zM2v\\n/lnfdfiHzs/EFXh3j86hXB01j4S+UI9bT2ImG7/umc0GCJja+yih7FGOmGjhWoT/\\nVnCZ8d1dSUC59fC/rXqs0suArQqS/aNvw7h7TniO44jjFnIdt72RoRj590y4E3e1\\nh9tdK+J+lDrTMe5Tz3Ocy52+Kll/G9H21dyHBDnjz4avFtFOhtwcj2GCiL3qLuv6\\nF21NVh2RImltPNVVCrlMx2D4D5hqMDd+RdLBQwIDAQABAoIBAQCuBYB3zvZ7BAKa\\nnkn/usFQlugfqiw4QznLlKvwfXN7YEADZHDV5r3Td9w551UcK6SkG3XkN7OQEAiI\\nXj9LKVHjuplGr8fI0ZriNGGBFcgrpN42FNdfMj34xDE3g9Ut5FIlT02eKWwEhxpY\\nqc/94rEanlmWhQX2tE3vdsT4uk8bCDjnuip545MElS8RgqB4PrIpPDjfMyL05O7+\\nuhxjmzaHC6N0JpyostUUcilpy9dR3uIFjqr3pHHxW8d3KQEoaCbnE/y1o3vyE68n\\nFzCbrrmL0JYL8lt3o1wJd7H6bXiTNfkhWT31nHfa9H6OrpFmY2hjAFgjUXqga4aT\\nFEf+IYGBAoGBAP2JaMNc3XwDRbYvCTzSJNK18Hvgogo2jbmkAAWwtdEze1VaH1i5\\ngIG9TKYrRqYZCKanSbMJNfQhJXaPRQZweWl2l5XYNu0xHwqTfNRfToi6TctTYdsG\\nJpT6PBfMQ3gmD41hULGgJZtE9rHkYysV2WHUzt1riulNP7wLuFZEVN9TAoGBAO75\\nMWM+q0ty6mx69Z/OA6xTEpWcWp/AatnMVyx2Cbq7dVwMTUt3ZyJYteiyrNuVYzil\\n3r57a1NLI2jyXSMsIBFCmGOiliHCFhb5j9mqjqVCMndSQfyUFXHN+HboDY5H/6nh\\ngVyTGtQtLcqiUadBVHwG+GwRF0j+4GOFBVAn3IhRAoGANu0r6xcVTKaQ1bU+9RQ5\\nsgSEgvHZaQkEjzE+Z/0uhpH2qDkMDXaypmjSUDWT6qwFucUILGdUbv03VoYXWBvJ\\n0dVFZcsfqyc3KF1nb8LuCqUB8Yiw7Hhap2/G0QkJnN/yApkQmEqvmwHHyCZIba8g\\nFxe5cTK5N23q3txTJS1Mjc0CgYEA6oR8VjP7MHje24SzpZcsfaGsvEwLl3jio0jo\\nR8ay96DdpG6iMCxGdyttoaYvAvMQw1fsRyyhf8VeQ7CNfGXiNm4PVMAK66gnu3ZA\\nGFFL2ZwkIHsGLNQ6U0vZyP8jxnybxuB57hqjGpQXuyMeVLBFM843Xv5KxqTyGlhB\\nhY8I2nECgYAvTTxOiu8ECMAeWNprJsjBJ1Ok1MkjcNYpjflzXSZSvS1Y4mTBqvaT\\n+X/uuZK59j4Rnhau4TZPLUmQ8kJuryIQa5o6jASM75lwTorEEtciIeDUcLLmKc5g\\ndrSpTYRMJoC9A9Hs/9X/HoBGxZ083VU1jH9FqT8vl1vBSFv1pBBuGg==\\n-----END RSA PRIVATE KEY-----', active_branch='main', github_base_branch='main'), mode='create_review_request')]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "llm = ChatOpenAI(temperature=0, model='gpt-4')\n",
    "github = GitHubAPIWrapper()\n",
    "tools = GitHubToolkit.from_github_api_wrapper(github).get_tools()\n",
    "agent = initialize_agent(\n",
    "    tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True\n",
    ")\n",
    "print(\"Available tools:\")\n",
    "tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools[1].args"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import langchain\n",
    "langchain.debug = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agent.run(\n",
    "    \"You have the software engineering capabilities of a Google Principle engineer. You are tasked with completing the most recent issue in the provided github repo. Please read that issue and complete it. Make sure to create a new branch before making changes, then open a PR and create a review request for the user who opened the issue.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3mFirst, I need to use the \"Search code\" tool to find files and code snippets related to \"RNA-seq\" and \"fastq\".\n",
      "Action: Search code\n",
      "Action Input: RNA-seq fastq\u001b[0m\n",
      "Observation: \u001b[33;1m\u001b[1;3mShowing top 2 of 2 results:\n",
      "Filepath: `README.md`\n",
      "File contents: # ML4Bio\n",
      "LLMs to execute Bioinformatics workflows, esp. RNA-seq\n",
      "\n",
      "Bio Tools Documentation:\n",
      "\n",
      "Starter packages we have docs for: `bioconductor, fastqc, multiqc, trimmomatic, STAR, gffread, samtools, salmon, DESeq2 and ggpubr`\n",
      "\n",
      "Completed:\n",
      "* https://www.uiuc.chat/ml4bio-fastqc\n",
      "* https://www.uiuc.chat/ml4bio-bioconductor\n",
      "* https://www.uiuc.chat/ml4bio-multiqc\n",
      "* https://www.uiuc.chat/ml4bio-star\n",
      "\n",
      "<END OF FILE>\n",
      "Filepath: `Report_WholeBrain/Report_Antonson_WholeBrain_2022Mar.Rmd`\n",
      "File contents: ---\n",
      "title: \"Antonson RNA-Seq Preliminary Report, WholeBrain\"\n",
      "author: \"Jessica Holmes and Jenny Drnevich, HPCBio, University of Illinois\"\n",
      "date: \"`r format(Sys.time(), '%B %d,  %Y')`\"\n",
      "output:\n",
      "  html_document:\n",
      "    df_print: paged\n",
      "    toc: yes\n",
      "    self_contained: yes\n",
      "  word_document:\n",
      "    toc: yes\n",
      "  pdf_document:\n",
      "    toc: yes\n",
      "params:\n",
      "  filebasename: Report_Antonson_2022Mar\n",
      "  anno: 109\n",
      "  anno_url: https://www.ncbi.nlm.nih.gov/genome/annotation_euk/Mus_musculus/109/\n",
      "  box_url: https://uofi.box.com/s/6ncwee31vmjs7huh6nfxlu9to3e261zz\n",
      "  fastqc_ver: 0.11.8\n",
      "  fc_ver: 2.0.0\n",
      "  ftp_url: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.27_GRCm39/\n",
      "  genome: GRCm39 (Mouse) ; A/X-31(H3N2)\n",
      "  genome_host: NCBI\n",
      "  genome_host_url: https://www.ncbi.nlm.nih.gov/genome/?term=mus+musculus ; https://www.ncbi.nlm.nih.gov/nuccore/?term=txid132504[Organism:noexp\n",
      "  genome_file: GCF_000001635.27_GRCm39_genomic.fna.gz ; influenza_A_X-31_H3N2_genes.fa \n",
      "  gtf_file: GCF_000001635.27_GRCm39_genomic.gff.gz ; influenza_A_X-31_H3N2.gff3\n",
      "  host: NCBI\n",
      "  multiqc_ver: 1.9\n",
      "  multiqc_report: Project_WB_RNASeq_multiqc_report.htm\n",
      "  salmon_ver: 1.4.0\n",
      "  species: Mus musculus ; H3N2\n",
      "  star_ver: 2.7.10a\n",
      "  transcriptome: 109\n",
      "  transcriptome_file: GCF_000001635.27_GRCm39_rna.fna.gz ; influenza_A_X-31_H3N2_rna.fa\n",
      "  trim_ver: 0.38\n",
      "---\n",
      "\n",
      "```{r setup, echo = FALSE, include=FALSE, warning=FALSE, message = FALSE, results='hide'}\n",
      "#change the path below to your current directory to run the file\n",
      "#knitr::opts_knit$set(root.dir =\"G:/My Drive/HPCBio/RNA-Seq/projects/ernelson/2019Feb-RNASeq/\") R markdown and all the files are in Final_Report, so no need to set working directory. \n",
      "knitr::opts_chunk$set(cache=TRUE, cache.lazy = FALSE)\n",
      "# knitr::knit_hooks$set(inline = function(x) {\n",
      "#   if(is.numeric(x)) scales::scales::comma(x)\n",
      "#   else x\n",
      "# })\n",
      "library(kableExtra)\n",
      "\n",
      "```\n",
      "\n",
      "```{r main codes, echo=FALSE, include=FALSE, warning=FALSE, message = FALSE, results='hide', cache=TRUE}\n",
      "library(ggplot2)\n",
      "library(dplyr)\n",
      "library(edgeR)\n",
      "library(gplots)\n",
      "library(rtracklayer)\n",
      "library(Glimma)\n",
      "library(tximport)\n",
      "library(org.Mm.eg.db)\n",
      "library(WGCNA)\n",
      "library(EDASeq)\n",
      "library(RUVSeq)\n",
      "\n",
      "\n",
      "#Read in sample information\n",
      "\n",
      "targets <- readTargets(\"Targets_Final.txt\")\n",
      "\n",
      "\n",
      "#cut down to WholeBrain only\n",
      "\n",
      "targets <- targets[targets$Sample.type == \"Whole brain\",]\n",
      "\n",
      "\n",
      "#Make better Trt and Labels\n",
      "\n",
      "targets$Trt <- substring(targets$Treatment, 1, 3) %>% factor(levels = c(\"Veh\",\"Flu\"))\n",
      "# targets$Trt <- \"Flu\"\n",
      "# targets$Trt[sample(1:10, 4)] <- \"Veh\"\n",
      "# targets$Trt <- factor(targets$Trt, levels = c(\"Veh\",\"Flu\"))\n",
      "targets$Label <- paste(targets$Trt, targets$Sample, sep = \"_\")\n",
      "\n",
      "\n",
      "\n",
      "#Add colors and sort\n",
      "targets$col <- dittoSeq::dittoColors()[as.numeric(targets$Trt)]\n",
      "targets <- targets[order(targets$Trt),]\n",
      "\n",
      "\n",
      "\n",
      "#load in transcript-level data\n",
      "\n",
      "load(\"SalmonSummarizedOutput.RData\")\n",
      "\n",
      "#If they have X added because they started with a number - remove them\n",
      "rownames(meta_info) <- gsub(\"X\", \"\",rownames(meta_info))\n",
      "\n",
      "\n",
      "#match the order\n",
      "temp <- match(as.character(targets$Sample), rownames(meta_info))\n",
      "meta_info <- meta_info[temp,]\n",
      "tx.all$abundance <- tx.all$abundance[,temp]\n",
      "tx.all$counts <- tx.all$counts[,temp]\n",
      "tx.all$length <- tx.all$length[,temp]\n",
      "\n",
      "\n",
      "#replace the old name \n",
      "rownames(meta_info) <- targets$Label\n",
      "colnames(tx.all$abundance) <- targets$Label\n",
      "colnames(tx.all$counts) <- targets$Label\n",
      "colnames(tx.all$length) <- targets$Label\n",
      "\n",
      "\n",
      "#do read fate plot\n",
      "targets$Total <- targets$NumReads\n",
      "ReadFate <- data.frame(#QCfiltered = targets$Total - meta_info$num_processed,\n",
      "  mapped = meta_info$num_mapped) / targets$Total * 100\n",
      "\n",
      "ReadFate$Label <- factor(targets$Label, levels = unique(targets$Label))\n",
      "ReadFate$Trt <- targets$Trt\n",
      "\n",
      "\n",
      "# Load transcript - gene mappings\n",
      "all_tx_info <- read.delim(\"trx_EGids_combined.txt\")\n",
      "\n",
      "names(all_tx_info)[3] <- c(\"symbol\")\n",
      "\n",
      "\n",
      "#get gene-level counts\n",
      "gene.all <- summarizeToGene(tx.all, all_tx_info[,c(1,2)], countsFromAbundance = \"lengthScaledTPM\")\n",
      "\n",
      "\n",
      "#cut down all_tx_info to unique genes\n",
      "#use product from transcript with largest counts\n",
      "\n",
      "all_tx_info$sumCounts <- rowSums(tx.all$counts)[all_tx_info$transcript_id]\n",
      "all_tx_info$sumCounts[is.na(all_tx_info$sumCounts)] <- 0\n",
      "all_tx_info <- all_tx_info[order(all_tx_info$sumCounts, decreasing = TRUE),]\n",
      "\n",
      "temp <- all_tx_info[!duplicated(all_tx_info$gene_id),-1]\n",
      "temp <- temp[match(rownames(gene.all$counts), temp$gene_id),]\n",
      "rownames(temp) <- temp$gene_id\n",
      "temp$ENTREZID <- gsub(\"GeneID:\",\"\", temp$gene_id)\n",
      "\n",
      "\n",
      "#create dgelist\n",
      "d <- DGEList(gene.all$counts, samples = targets, genes = temp[,c(\"ENTREZID\",\"symbol\",\"product\")])\n",
      "\n",
      "\n",
      "#Do TMM norm ----\n",
      "\n",
      "d <- calcNormFactors(d)\n",
      "\n",
      "logCPM <- cpm(d, log=T)\n",
      "\n",
      "\n",
      "\n",
      "####Filtering####\n",
      "\n",
      "min_cpm <- 1\n",
      "min_samp <- 3\n",
      "\n",
      "i.filter <- rowSums(logCPM > log2(min_cpm)) >= min_samp\n",
      "\n",
      "\n",
      "#cut down to filtered gene list\n",
      "d.filt <- d[i.filter, , keep.lib.sizes=F]\n",
      "\n",
      "#re-do TMM factors, just in case:\n",
      "d.filt <- calcNormFactors(d.filt)\n",
      "\n",
      "\n",
      "\n",
      "#do final normalization----\n",
      "#default prior.count is now 2, use that\n",
      "logCPM.filt <- cpm(d.filt, log = T, prior.count = 2)\n",
      "\n",
      "\n",
      "#Do glMDSPlot\n",
      "\n",
      "\n",
      "temp <- glMDSPlot(logCPM.filt, top = 5000, labels = d.filt$samples$Label,\n",
      "                  groups = d.filt$samples[,c(\"Trt\", \"Sample\")], launch = FALSE,\n",
      "          html = \"old_MDSclustering_WholeBrain\", folder = \"interactive_plots\")\n",
      "\n",
      "plot.values <- as.data.frame(temp$points)\n",
      "colnames(plot.values) <- paste0(\"Dim\", 1:8)\n",
      "plot.values$var.exp <- round(temp$eig*100/sum(temp$eig),1)\n",
      "plot.values$Trt <- factor(d.filt$samples$Trt)\n",
      "plot.values$Label <- factor(d.filt$samples$Label)\n",
      "\n",
      "\n",
      "\n",
      "glimmaMDS(d.filt, top = 5000, labels = d$samples$Label,\n",
      "          html = \"MDSclustering_WholeBrain.html\")\n",
      "\n",
      "\n",
      "#deg ----\n",
      "Group <- factor(targets$Trt, levels = unique(targets$Trt))\n",
      "design0 <- model.matrix(~ Group)\n",
      "colnames(design0)[-1] <- paste0(levels(Group)[-1], \".vs.\", levels(Group)[1])\n",
      "\n",
      "\n",
      "#put normalized values in EList object type\n",
      "e <- new(\"EList\", list(E =logCPM.filt, genes=d.filt$genes, targets=targets))\n",
      "\n",
      "#fit model ----\n",
      "fit0 <- lmFit(e, design = design0) %>% eBayes(trend = TRUE)\n",
      "\n",
      "\n",
      "# codes_sep05 <- decideTests(fit0)\n",
      "# codes_glob05 <- decideTests(fit0, method = \"global\")\n",
      "# codes_glob1 <- decideTests(fit0, method = \"global\", p.value = 0.1)\n",
      "\n",
      "\n",
      "#get results \n",
      "res.temp <- topTable(fit0, coef = 2, n = Inf, sort.by = \"none\")\n",
      "\n",
      "\n",
      "#Make objects for heatmap\n",
      "\n",
      "col.pan <- colorpanel(100, \"blue\", \"white\", \"red\")\n",
      "heat.h1 <- t(scale(t(logCPM.filt[res.temp$P.Value < 0.05,])))\n",
      "x.cluster.h1 <- hclust(dist(heat.h1))\n",
      "\n",
      "rowCols.h1 <- cutreeStaticColor(x.cluster.h1, cutHeight = 5.1, minSize = 1 )\n",
      "\n",
      "heatVal.h1 <- data.frame(ENTREZID=gsub(\"GeneID:\", \"\", rownames(heat.h1)),\n",
      "                         heat.h1,\n",
      "                         HeatmapColor=rowCols.h1)\n",
      "heatVal.h1 <- heatVal.h1[rev(x.cluster.h1$order),]\n",
      "heatVal.h1$TopToBottom <- 1:nrow(heatVal.h1)\n",
      "\n",
      "\n",
      "\n",
      "# get results\n",
      "# \n",
      "results.out0 <- res.temp[,1:8]\n",
      "results.out0$t <- 2^abs(results.out0$logFC) * sign(results.out0$logFC)\n",
      "names(results.out0)[6] <- c(\"regFC\")\n",
      "\n",
      "\n",
      "#merge gene position in heatmap\n",
      "results.out0  <-  results.out0 %>% left_join(heatVal.h1 %>% dplyr::select(ENTREZID, HeatmapColor, TopToBottom))\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "source(\"getGO.R\")\n",
      "source(\"getPath.R\")\n",
      "\n",
      "#get GO terms\n",
      "tempGO <- getGO(org.Mm.eg.db, keys = results.out0$ENTREZID,keytype = \"ENTREZID\")\n",
      "\n",
      "results.out0 <-  left_join(results.out0, tempGO)\n",
      "\n",
      "#get KEGG pathways\n",
      "tempKEGG <- getPath(keys = results.out0$ENTREZID, updatePath = TRUE, species = \"mmu\")\n",
      "\n",
      "results.out <-  left_join(results.out0, tempKEGG)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "#order columns\n",
      "results.out <- results.out %>% dplyr::select(ENTREZID:product, HeatmapColor, TopToBottom, everything())\n",
      "\n",
      "write.table(results.out, file = \"Antonson_WholeBrain_GeneResults_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "#write out raw counts\n",
      "\n",
      "count.out <- cbind(d$genes, kept = i.filter,d$counts)\n",
      "\n",
      "write.table(count.out, file = \"Antonson_WholeBrain_Gene_level_counts_2022-05-12.txt\", sep = \"\\t\", row.names = FALSE)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "# \n",
      "# \n",
      "# ####make interactive results table----\n",
      "# \n",
      "# \n",
      "#write interactive results\n",
      "rawPcodes <- decideTests(fit0, adjust.method = \"none\")\n",
      "\n",
      "glMDPlot(fit0, coef = 2, counts = e$E,\n",
      "           groups = Group,\n",
      "           side.main = \"symbol\",\n",
      "           samples = targets$Label,\n",
      "           status = rawPcodes,\n",
      "           main = \"WholeBrain, Flu vs. Veh, raw P < 0.05\",\n",
      "           sample.cols = targets$col, folder = \"interactive_plots/\",\n",
      "           html = \"old_WholeBrain_rawP05\", launch = FALSE,\n",
      "         p.adj.method = \"none\")\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "#Try RUV with increasing numbers of k to find best number ----\n",
      "\n",
      "num0.05 <- sum(res.temp$P.Value < 0.05)\n",
      "diff <- makeGroups(Group)\n",
      "\n",
      "for(k in 1:6) {\n",
      "  wobj <- RUVs(e$E, cIdx = res.temp$P.Value > 0.5, k = k, diff, isLog = TRUE)\n",
      "  #add factors to model\n",
      "  model.ruv <- cbind(design0, wobj$W)\n",
      "  fit.ruv <- lmFit(e, model.ruv) %>% eBayes( trend=TRUE)\n",
      "  anova.ruv <- topTable(fit.ruv, coef = 2, number = Inf, sort.by = \"none\")\n",
      "  num0.05 <- c(num0.05, sum(anova.ruv$P.Value < 0.05))\n",
      "}\n",
      "\n",
      "num0.05 <- data.frame(NumberOfFactors = 0:k, NumberSig05 = num0.05)\n",
      "num0.05$NumberOfFactors <- factor(num0.05$NumberOfFactors )\n",
      "\n",
      "\n",
      "\n",
      "k <- 4\n",
      "wobj_final <- RUVs(e$E, cIdx = res.temp$P.Value > 0.5, k = k, diff, isLog = TRUE)\n",
      "\n",
      "\n",
      "\n",
      "temp <- glMDSPlot(wobj_final$normalizedCounts, top = 5000, labels = d.filt$samples$Label, \n",
      "                  groups = d.filt$samples[,c(\"Trt\")],\n",
      "                  html = \"old_MDSclustering_WholeBrain_RUV\", folder = \"interactive_plots/\", launch = FALSE)\n",
      "\n",
      "plot.values4 <- as.data.frame(temp$points)\n",
      "colnames(plot.values4) <- paste0(\"Dim\", 1:8)\n",
      "plot.values4$var.exp <- round(temp$eig*100/sum(temp$eig),1)\n",
      "plot.values4$Trt <- factor(d.filt$samples$Trt)\n",
      "plot.values4$Label <- factor(d.filt$samples$Label)\n",
      "\n",
      "\n",
      "#Fit the model\n",
      "\n",
      "\n",
      "design.ruv4 <- cbind(design0, wobj_final$W)\n",
      "fit.ruv4 <- lmFit(e, design = design.ruv4) %>% eBayes(trend = T )\n",
      "\n",
      "#get RUV results\n",
      "\n",
      "cont.matrix4 <- makeContrasts(Flu.vs.Veh,\n",
      "                              levels = design.ruv4)\n",
      "\n",
      "\n",
      "fit.ruv4b  <- eBayes(contrasts.fit(fit.ruv4, cont.matrix4), trend = TRUE)\n",
      "\n",
      "\n",
      "# \n",
      "# \n",
      "# \n",
      "\n",
      "codes_fdr0.1_ruv <- decideTests(fit.ruv4b, p.value = 0.1)\n",
      "\n",
      "\n",
      "results.out <- topTable(fit.ruv4b, coef = 1, num = Inf, sort.by = \"none\")\n",
      "results.out$t <- 2^abs(results.out$logFC) * sign(results.out$logFC)\n",
      "names(results.out)[6] <- c(\"regFC\")\n",
      "\n",
      "###Heatmaps----\n",
      "sig.index.int <- results.out$adj.P.Val < 0.1\n",
      "\n",
      "col.pan <- colorpanel(100, \"blue\", \"white\", \"red\")\n",
      "heat.h2 <- t(scale(t(wobj_final$normalizedCounts[sig.index.int,])))\n",
      "x.cluster.h2 <- hclust(dist(heat.h2))\n",
      "\n",
      "\n",
      "rowCols.h2 <- cutreeStaticColor(x.cluster.h2, cutHeight = 5.6, minSize = 1 )\n",
      "\n",
      "#output gene position in the heatmap\n",
      "heatVal.h2 <- data.frame(ENTREZID=gsub(\"GeneID:\", \"\", rownames(heat.h2)), \n",
      "                         heat.h2,\n",
      "                         HeatmapColor=rowCols.h2)\n",
      "heatVal.h2 <- heatVal.h2[rev(x.cluster.h2$order),]\n",
      "heatVal.h2$TopToBottom <- 1:nrow(heatVal.h2)\n",
      "\n",
      "#merge gene position in heatmap\n",
      "results.out  <-  results.out %>% left_join(heatVal.h2 %>% dplyr::select(ENTREZID, HeatmapColor, TopToBottom))\n",
      "\n",
      "\n",
      "#Add in GO/KEGG\n",
      "\n",
      "\n",
      "#get GO terms\n",
      "tempGO <- getGO(org.Mm.eg.db, keys = results.out$ENTREZID,keytype = \"ENTREZID\")\n",
      "\n",
      "results.out <-  left_join(results.out, tempGO)\n",
      "\n",
      "#get KEGG pathways\n",
      "tempKEGG <- getPath(keys = results.out$ENTREZID, updatePath = TRUE, species = \"mmu\")\n",
      "\n",
      "results.out <-  left_join(results.out, tempKEGG)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "#order columns\n",
      "results.out <- results.out %>% dplyr::select(ENTREZID:product, HeatmapColor, TopToBottom, everything())\n",
      "\n",
      "\n",
      "#Write out RUV values\n",
      "\n",
      "temp <- cbind(d.filt$samples[,c(4, 9, 11, 2, 3, 7, 8)], wobj_final$W)\n",
      "names(temp)[8:11] <- paste0(\"RUV_\", 1:k)\n",
      "write.csv(temp, file = \"SampleInfo_RUVvariables_WholeBrain_2022-05-12.csv\", row.names = FALSE)\n",
      "\n",
      "\n",
      "#write out regular results\n",
      "\n",
      "#write un-corrected and corrected expression values\n",
      "cpm.out <- cbind(results.out[,1:3], logCPM.filt)\n",
      "RUV.out <- cbind(results.out[,1:3], wobj_final$normalizedCounts)\n",
      "\n",
      "write.table(cpm.out, file = \"Antonson_WholeBrain_logCPMvalues_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "write.table(RUV.out, file = \"Antonson_WholeBrain_logCPMValues_RUVcorrected_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "\n",
      "#write out full results\n",
      "\n",
      "write.table(results.out, \"Antonson_WholeBrain_GeneResults_RUV_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "\n",
      "#write interactive results\n",
      "glMDPlot(fit.ruv4b, coef = 1, counts = wobj_final$normalizedCounts,\n",
      "           groups = Group,\n",
      "           side.main = \"symbol\",\n",
      "           samples = targets$Label,\n",
      "           status = codes_fdr0.1_ruv,\n",
      "           main = \"WholeBrain, Flu vs. Veh RUV, FDR P < 0.1\",\n",
      "           sample.cols = targets$col, folder = \"interactive_plots/\",\n",
      "           html = \"old_WholeBrain_RUV_FDR0.1\", launch = FALSE)\n",
      "\n",
      "\n",
      "\n",
      "#Do quick go testing for both ----\n",
      "\n",
      "up <- results.out0$ENTREZID[results.out0$P.Value < 0.05 & results.out0$logFC > 0]\n",
      "go.up <- goana(de = up, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "go.up <- go.up[go.up$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getGO(org.Mm.eg.db, keys = up, keytype = \"ENTREZID\",\n",
      "              GOALL = TRUE, summarizeBy = \"go\", EGorSYM = \"SYM\")\n",
      "#Add in DE symbols\n",
      "go.up <- tibble::rownames_to_column(go.up, \"GOID\") %>% left_join(temp)\n",
      "go.up <- go.up[order(go.up$P.DE),]\n",
      "write.table(go.up, file = \"GO_up_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "kegg.up <- kegga(de = up, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "kegg.up <- kegg.up[kegg.up$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getPath(org.Mm.eg.db, keys = up, keytype = \"ENTREZID\",\n",
      "              updatePath = TRUE, summarizeBy = \"pathway\", EGorSYM = \"SYM\",\n",
      "              keepSpecies = TRUE, species = \"mmu\")\n",
      "#Add in DE symbols\n",
      "kegg.up <- tibble::rownames_to_column(kegg.up, \"PATHID\") %>% left_join(temp)\n",
      "kegg.up <- kegg.up[order(kegg.up$P.DE),]\n",
      "write.table(kegg.up, file = \"KEGG_up_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "\n",
      "\n",
      "down <- results.out0$ENTREZID[results.out0$P.Value < 0.05 & results.out0$logFC < 0]\n",
      "go.down <- goana(de = down, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "go.down <- go.down[go.down$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getGO(org.Mm.eg.db, keys = down, keytype = \"ENTREZID\",\n",
      "              GOALL = TRUE, summarizeBy = \"go\", EGorSYM = \"SYM\")\n",
      "#Add in DE symbols\n",
      "go.down <- tibble::rownames_to_column(go.down, \"GOID\") %>% left_join(temp)\n",
      "go.down <- go.down[order(go.down$P.DE),]\n",
      "write.table(go.down, file = \"GO_down_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "kegg.down <- kegga(de = down, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "kegg.down <- kegg.down[kegg.down$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getPath(org.Mm.eg.db, keys = down, keytype = \"ENTREZID\",\n",
      "              updatePath = TRUE, summarizeBy = \"pathway\", EGorSYM = \"SYM\",\n",
      "              keepSpecies = TRUE, species = \"mmu\")\n",
      "#Add in DE symbols\n",
      "kegg.down <- tibble::rownames_to_column(kegg.down, \"PATHID\") %>% left_join(temp)\n",
      "kegg.down <- kegg.down[order(kegg.down$P.DE),]\n",
      "write.table(kegg.down, file = \"KEGG_down_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "#Do for RUV results\n",
      "\n",
      "up <- results.out0$ENTREZID[results.out0$P.Value < 0.05 & results.out0$logFC > 0]\n",
      "go.up <- goana(de = up, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "go.up <- go.up[go.up$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getGO(org.Mm.eg.db, keys = up, keytype = \"ENTREZID\",\n",
      "              GOALL = TRUE, summarizeBy = \"go\", EGorSYM = \"SYM\")\n",
      "#Add in DE symbols\n",
      "go.up <- tibble::rownames_to_column(go.up, \"GOID\") %>% left_join(temp)\n",
      "go.up <- go.up[order(go.up$P.DE),]\n",
      "write.table(go.up, file = \"GO_up_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "kegg.up <- kegga(de = up, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "kegg.up <- kegg.up[kegg.up$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getPath(org.Mm.eg.db, keys = up, keytype = \"ENTREZID\",\n",
      "              updatePath = TRUE, summarizeBy = \"pathway\", EGorSYM = \"SYM\",\n",
      "              keepSpecies = TRUE, species = \"mmu\")\n",
      "#Add in DE symbols\n",
      "kegg.up <- tibble::rownames_to_column(kegg.up, \"PATHID\") %>% left_join(temp)\n",
      "kegg.up <- kegg.up[order(kegg.up$P.DE),]\n",
      "write.table(kegg.up, file = \"KEGG_up_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "\n",
      "\n",
      "down <- results.out0$ENTREZID[results.out0$P.Value < 0.05 & results.out0$logFC < 0]\n",
      "go.down <- goana(de = down, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "go.down <- go.down[go.down$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getGO(org.Mm.eg.db, keys = down, keytype = \"ENTREZID\",\n",
      "              GOALL = TRUE, summarizeBy = \"go\", EGorSYM = \"SYM\")\n",
      "#Add in DE symbols\n",
      "go.down <- tibble::rownames_to_column(go.down, \"GOID\") %>% left_join(temp)\n",
      "go.down <- go.down[order(go.down$P.DE),]\n",
      "write.table(go.down, file = \"GO_down_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "kegg.down <- kegga(de = down, universe = results.out0$ENTREZID, species = \"Mm\")\n",
      "#remove terms without at least 2 DE genes\n",
      "kegg.down <- kegg.down[kegg.down$DE > 1,]\n",
      "#Get gene symbols\n",
      "temp <- getPath(org.Mm.eg.db, keys = down, keytype = \"ENTREZID\",\n",
      "              updatePath = TRUE, summarizeBy = \"pathway\", EGorSYM = \"SYM\",\n",
      "              keepSpecies = TRUE, species = \"mmu\")\n",
      "#Add in DE symbols\n",
      "kegg.down <- tibble::rownames_to_column(kegg.down, \"PATHID\") %>% left_join(temp)\n",
      "kegg.down <- kegg.down[order(kegg.down$P.DE),]\n",
      "write.table(kegg.down, file = \"KEGG_down_rawP05_2022-05-12.txt\",\n",
      "            row.names = FALSE, sep = \"\\t\")\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "```\n",
      "\n",
      "## Location of results and codes for reproducibility\n",
      "\n",
      "All deliverables for the basic analysis are in a zipped file that can be [downloaded from box.com](`r params$box_url`). Please unzip the file after transferring to your computer (PCs: right click and select \"Extract all\"; Macs: double-click to extract). Folders and Files:\n",
      "\n",
      "1. **interactive_plots** folder. Contains non-stand alone interactive MDS clustering and results .html files, which must be kept with the ccs/ and js/ sub-folders to work.\n",
      "\n",
      "2. **Report_Antonson_WholeBrain_2022Mar_files** folder. Contains separate .png files of the figures in this report.\n",
      "\n",
      "3. **Antonson_WholeBrain_Gene_level_counts_2022-05-12.xlsx**. Contains gene counts (summed from length-normalized transcript counts) for all genes and a column showing which were kept after filtering.\n",
      "\n",
      "4. **Antonson_WholeBrain_GeneResults_2022-05-12.xlsx**. Contains gene-level statistical results for Flu vs. Veh; see the tab at the bottom with column explanations. \n",
      "\n",
      "5. **Antonson_WholeBrain_GeneResults_RUV_2022-05-12.xlsx**. Contains gene-level statistical results for Flu vs. Veh from the model including RUV factors; see the tab at the bottom with column explanations. \n",
      "\n",
      "6. **Antonson_WholeBrain_logCPMvalues_2022-05-12.xlsx**. Normalized logCPM values for kept genes for each sample; these were used as the input to both statistical models and for visualizations (MDS clustering, heatmap, top right plot of interactive results) of the plain Flu vs. Veh comparison.\n",
      "\n",
      "7. **Antonson_WholeBrain_logCPMvalues_RUVcorrected_2022-05-12.xlsx**. Normalized logCPM values that have had the effects of the RUV factors; these were used only used for visualizations (MDS clustering, heatmap, top right plot of interactive results) of the Flu vs. Veh + RUV factors comparison.\n",
      "\n",
      "8. **MDSclustering_WholeBrain.html**. Stand-alone version of the new interactive MDS clustering plot.\n",
      "\n",
      "9. **MDSclustering_WholeBrain_RUVremoved.html**. Stand-alone version of the new interactive MDS clustering plot after RUV removal.\n",
      "\n",
      "10. **Report_Antonson_Microglia_2022Mar.Rmd**. R markdown file with the R codes for the entire analysis and to make this html report.\n",
      "\n",
      "11. **Report_Antonson_Microglia_2022Mar.html**. This html report.\n",
      "\n",
      "12. **SalmonSummarizedOutput.RData**. Transcript-level counts and other output from Salmon, in RData format to be read into R using Report_Antonson_Microglia_2022Mar.Rmd.\n",
      "\n",
      "10. **SampleInfo_RUVvariables_WholeBrain_2022-05-12.csv**. Sample information including TMM normalization values and RUV factors.\n",
      "\n",
      "8. **WholeBrain_rawP05.html**. Stand-alone version of the interactive results for the plain Flu vs. Veh comparison.\n",
      "\n",
      "9. **WholeBrain_RUV_FDR0.1.html**. Stand-alone version of the interactive results for the Flu vs. Veh + RUV factors comparison.\n",
      "\n",
      "7. getGO.R, getPath.R, Targets_Final.txt, trx_EGids_combined.txt. Other input files needed to run Report_Antonson_Microglia_2022Mar.Rmd codes.\n",
      "\n",
      "\n",
      "\n",
      "## Quality Check, alignment and count generation\n",
      "\n",
      "### Quality check on the raw data\n",
      "\n",
      "We reviewed the QC report, \"`r params$multiqc_report`\" sent by the sequencing center that performed [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)[^fastqc] (version `r params$fastqc_ver`) on individual samples then was summarized into a single html report by using [MultiQC](http://multiqc.info/)[^multiqc] version `r params$multiqc_ver`. Average per-base read quality scores are over 30 in all samples and no adapter sequences were found indicating those reads are high in quality. Thus, we skipped the trimming step and directly proceed to transcripts mapping and quantification.\n",
      "\n",
      "### Reference sequences\n",
      "\n",
      "All reference files were downloaded from [NCBI's ftp site](`r params$ftp_url`). The *`r params$species`* transcriptome file \"`r params$transcriptome_file`\" from Annotation [`r params$anno`](`r params$anno_url`) from [`r params$genome_host`](`r params$genome_host_url`) was used for quasi-mapping and count generation. This transcriptome is derived from genome `r params$genome`; Since the quasi-mapping step only uses transcript sequences, the gene model file \"`r params$gtf_file`\" was solely used to generate transcript-gene mapping table (\"trx_EGids_GRCm39_annot109.txt\") for obtaining gene-level counts. [NOTE - NEED TO MODIFY THE ABOVE DESCRIPTION BECAUSE OF THE INCLUSION OF THE VIRAL SEQUENCES]\n",
      "\n",
      "\n",
      "### Alignment and gene-level quantification\n",
      "\n",
      "[Salmon](https://combine-lab.github.io/salmon/)[^salmon] version `r params$salmon_ver` was used to quasi-map reads to the transcriptome and quantify the abundance of each transcript. The transcriptome was first indexed using the decoy-aware method in Salmon with the entire genome file \"`r params$genome_file`\" as the decoy sequence. Then quasi-mapping was performed to map reads to transcriptome with additional arguments `--seqBias` and `--gcBias` to correct sequence-specific and GC content biases, `--numBootstraps=30` to compute bootstrap transcript abundance estimates and `--validateMappings` to help improve the accuracy of mappings. Gene-level counts were then estimated based on transcript-level counts using the [\"bias corrected counts without an offset\"](http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#use-with-downstream-bioconductor-dge-packages) method from the [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html) package. This method provides more accurate gene-level counts estimates and keeps multi-mapped reads in the analysis compared to traditional alignment-based method[^tximport].\n",
      "\n",
      "Percentage of reads mapped to the transcriptome ranged from `r round(min(ReadFate$mapped),1)` to `r round(max(ReadFate$mapped),1)`% (Figure 1), which is reasonable for a eukaryotic organism. There is no obvious difference in mapping rate between treatments. The unmapped reads were discarded and the number of remaining reads (range: `r round(min(d$samples$lib.size/1e6),1)` - `r round(max(d$samples$lib.size/1e6),1)` million per sample) were kept for further analysis. \n",
      "\n",
      "```{r Figure 1,fig.cap = \"Figure 1. Percentage of reads mapped to transcriptome\", echo = F, message=FALSE, fig.width = 10, fig.height = 6}\n",
      "ggplot(data=ReadFate, aes(x=Label, y=mapped, fill = Trt)) +\n",
      "  geom_bar(stat=\"identity\", position=position_dodge()) +\n",
      "  theme(axis.text.x = element_text(angle = 90, hjust = 1))+\n",
      "  ylab(\"Percentage\")+\n",
      "  scale_y_continuous(breaks = seq(0,100,by = 10), limits = c(0,100))+ \n",
      "  scale_fill_manual(values = dittoSeq::dittoColors()[1:5])\n",
      "```\n",
      "\n",
      "\n",
      "## Sample QC\n",
      "\n",
      "### Normalization \n",
      "\n",
      "When comparing expression levels, the numbers of reads per gene need to be normalized not only because of the differences in total number of reads, but because there could be differences in RNA composition such that the total number of reads would not be expected to be the same. The TMM (trimmed mean of M values) normalization[^tmm] in the [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) package[^edger2] uses the assumption of *most genes do not change* to calculate a normalization factor for each sample to adjust for such biases in RNA composition. In this dataset, TMM normalization factors fluctuates between `r round(min(d$samples$norm.factors),2)` and `r round(max(d$samples$norm.factors),2)`. A TMM value below 1 means the total library size is larger than it should be, and a TMM value above 1 means the total libary size is smaller than it should be. There is no obvious difference between Veh and Flu samples.\n",
      "\n",
      "```{r Figure 2, fig.cap = \"Figure 2. TMM nomalization factors to correct for RNA composition\", echo = F, message=FALSE, fig.width = 10, fig.height = 6}\n",
      "\n",
      "temp <- d$samples\n",
      "temp$Label <- factor(targets$Label, levels = unique(targets$Label))\n",
      "\n",
      "ggplot(temp, aes(y = norm.factors, x = Label, fill = Trt)) +\n",
      "  geom_bar(stat=\"identity\", position=position_dodge()) +\n",
      "  labs(x = \"Sample\", y = \"TMM normalization factor\") +\n",
      "  geom_hline(yintercept = 1, lty = 2) +\n",
      "  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + \n",
      "  scale_fill_manual(values = dittoSeq::dittoColors()[1:5])\n",
      "```\n",
      "\n",
      "### Filtering \n",
      "\n",
      "While the `r params$host` `r params$genome` Annotation [`r params$anno`](`r params$anno_url`) gene models have a total of `r scales::comma(nrow(d))` genes, many of these might not have detectable expression in these samples. Therefore, we set the detection threshold at `r min_cpm` cpm (counts per million) in at least `r min_samp` sample (because there was only 1 replicate of the low dose), which resulted in `r scales::comma(sum(!i.filter))` genes being filtered out, leaving `r scales::comma(sum(i.filter))` genes to be analyzed for differential expression that contain `r scales::percent(sum(d.filt$counts) / sum(d$counts), 0.01)` of the reads. After filtering, TMM normalization was performed again and normalized log2-based count per million values (logCPM) were calculated using edgeR's[^edger1] `cpm()` function with `prior.count = 2` to help stabilize fold-changes of extremely low expression genes.\n",
      "\n",
      "\n",
      "### Clustering\n",
      "\n",
      "Multidimensional scaling in the [limma](https://bioconductor.org/packages/release/bioc/html/limma.html)[^limma] package was used as a sample QC step to check for outliers or batch effects. The normalized logCPM values of the top most 5,000 variable genes were chosen to construct the multidimensional scaling plot. Dimension 1 accounts for `r plot.values$var.exp[1]`% of the expression variation and Dimension 2 explains `r plot.values$var.exp[2]`% of the expression variation. However, the treatment replicates do not cluster tightly together. A stand-alone interactive version of the MDS clustering plot is in file \"MDSclustering_WholeBrain.html\".\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "```{r Figure 3, fig.cap = \"Figure 3. Multidimensional Scaling on the top 5,000 most variable genes. Dimension 1 vs. 2\", echo = F, message=FALSE, fig.width = 7, fig.height = 6}\n",
      "ggplot(plot.values, aes(Dim1, Dim2, color = Trt))+\n",
      "  geom_point(aes(fill = Trt),  size = 3)+theme_bw()+\n",
      "  xlab(paste0(\"Dimension 1 \",\"(\",plot.values$var.exp[1],\"%)\")) +\n",
      "  ylab(paste0(\"Dimension 2 \",\"(\",plot.values$var.exp[2],\"%)\")) +\n",
      "  geom_text(aes(label = Trt),hjust=0.5, vjust=-0.6) + \n",
      "  scale_color_manual(values = dittoSeq::dittoColors()[1:4])\n",
      "```\n",
      "\n",
      "## Differential gene expression analysis \n",
      "\n",
      "Differential gene expression (DE) analysis was performed using the limma-trend method[^trend1] using a model of ~ treatment. The pairwise comparison between Flu and Vehicle as the control was made and a False Discovery Rate method[^fdr] (FDR) performed. However, the lowest FDR p-value for any gene was `r min(results.out0$adj.P.Val)`. This does not automatically mean that no genes were differentially expressed, just the number is not substantially larger than the number of false-positives that are theoretically expected. Instead, we can look at the genes with the lowest raw p-values as those with the \"most\" evidence of differential expression. Using a raw p-value threshold of 0.05, there are `r sum(rawPcodes[,2] == 1)` genes with higher expression in Flu vs. Veh and `r sum(rawPcodes[,2] == -1)` genes with lower expression in Flu vs. Veh. These genes are visualized in a heatmap (Figure 4). \n",
      "\n",
      "\n",
      "\n",
      "```{r Figure 4, fig.cap = \"Figure 4. Heatmap of 321 genes with Flu vs. Veh raw p-value < 0.05\", echo = F, message=FALSE, fig.width = 10, fig.height = 10, cache = TRUE}\n",
      "\n",
      "heatmap.2(heat.h1, col=col.pan, Rowv=as.dendrogram(x.cluster.h1),Colv=FALSE, scale=\"none\",\n",
      "          trace=\"none\", dendrogram=\"row\", cexRow=0.9, cexCol=1.5, labRow = \"\",\n",
      "          keysize = 0.2,lwid = c(1.4,4), lhei = c(1,4), key.par=list(mar = c(4,2,2,1)),\n",
      "          margins = c(9,2),density.info = \"none\", key.xlab = \"SD from mean\",\n",
      "          colsep =c(4), sepcolor=\"black\",\n",
      "          sepwidth=c(0.05,0.05), \n",
      "          #RowSideColors = rowCols.h1\n",
      "          )\n",
      "\n",
      "\n",
      "\n",
      "```\n",
      "\n",
      "\n",
      "### Interactive results\n",
      "\n",
      "Below is one example of the interactive results for the Flu vs. Veh pairwise comparison. The top left shows each gene as a point, with the average expression value on the x-axis and the log2(fold-change) on the y-axis. Genes with raw p-value < 0.05 are highlighted in red (up-regulated; status = 1) or blue (down-regulated; status = -1). Mousing over any point will pop up a box with information and clicking on a point will plot the individual sample values on the top right. At the bottom is a searchable/sortable table of the genes. Clicking on any gene in the table will highlight it in both top plots. A stand-alone interactive .html file is also in the main output directory. There is also non stand-alone .html files of similar interactive results available in the \"interactive_plots\" directory. This is an older version that are better in a few ways (quicker switching to top right plot and auto-scaling its y-axis) but do not contain all of the functionality of the newer interactive plots. In the old version the **AdjPValue** column actually gives the raw p-values whereas both raw and FDR corrected p-values are in the new versions. \n",
      "\n",
      "In the plot below, try:\n",
      "\n",
      "+ Clicking points in the summary plot or rows in the table to plot the gene expression of the selection.\n",
      "    + Clicking genes in the table after selecting individual points will remove the previous selection.\n",
      "+ Searching for individual genes using the search box. The search results are displayed in the table.\n",
      "    + If genes are currently selected, the search box will not function.\n",
      "+ Setting a maximum value for the y-axis of the expression plot using the max_y_axis field. \n",
      "    + This allows for comparison of gene expression between genes on a comparable scale.\n",
      "+ Saving the currently selected genes using the Save Data dropdown. \n",
      "    + From here, **you can also choose to save the entire table.**\n",
      "+ Saving the summary plot or expression plot in either PNG or SVG formats, using the \"Save Data\" dropdown.\n",
      "\n",
      "\n",
      "```{r interactive1, echo = FALSE, eval = TRUE}\n",
      "\n",
      "i <- 2\n",
      "\n",
      "#annot <- data.frame(results.out0[,c(2:3, 7)])\n",
      "glimmaMA(fit0, d.filt, counts = e$E, transform.counts = \"none\",\n",
      "         groups =d.filt$samples$Trt, coef = i,\n",
      "         status  = rawPcodes, main = colnames(rawPcodes)[i],\n",
      ")\n",
      "         #anno = annot)\n",
      "\n",
      "glimmaMA(fit0, d.filt, counts = e$E, transform.counts = \"none\",\n",
      "         groups =d.filt$samples$Trt, coef = i,\n",
      "         status  = rawPcodes, main = colnames(rawPcodes)[i],\n",
      "         html = \"WholeBrain_rawP05.html\")\n",
      "\n",
      "\n",
      "```\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "## Remove Unwanted Variation (RUV) Analysis\n",
      "\n",
      "### Iterative RUV\n",
      "\n",
      "\"Remove Unwanted Variation\"[^ruvseq] (RUV) is a method to estimate extra factors that can be added to the statistical model as you would known co-variates like age or weight, assuming that these extra factors are spurious technical variation in the samples and not biologically related. [Peixoto et al. (2015)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4652761/)[^iterRUV] show that extra normalization using RUV is often necessary in RNA-Seq experiments and helps to improve biological insights. They recommend to iteratively increase the number of factors until certain metrics become satisfactory. I counted the number of genes with raw p-value < 0.05 with increasing numbers of RUV factors (Figure 6). The number of genes increased across 2,3, and 4 RUV factors but the rate of increase slows between 4 and 5 and plateaus after 5. With only 10 samples I wanted to keep the number of RUV factors lower so I picked 4. The estimated RUV factor values for each sample are in the file  \"SampleInfo_RUVvariables_WholeBrain_2022-05-12.csv\". \n",
      "\n",
      "```{r Figure 6, fig.cap = \"Figure 6. Number of genes with oneway ANOVA FDR p-value < 0.05 using increasing numbers of RUV factors in the statistical model\", echo = F, message=FALSE, fig.width = 5, fig.height = 5}\n",
      "\n",
      "ggplot(data=num0.05, aes(x=NumberOfFactors, y=NumberSig05, group=1)) +\n",
      "  geom_line(linetype = \"dashed\", size = 1)+\n",
      "  geom_point(size = 3) +\n",
      "  xlab(\"Number of RUV factors\") + ylab(\"Number of genes\")\n",
      "  \n",
      "\n",
      "```\n",
      "\n",
      "\n",
      "\n",
      "### Clustering after RUV removal\n",
      "\n",
      "We can remove the effects of these extra factors from the normalized expression values in order to visualize the improvement in sample clustering (Figure 7). Dimension 1 mostly separates Veh from Flu samples and explains `r plot.values4$var.exp[1]`% of the variation in expression in the top 5,000 most variable genes. A stand-alone interactive version of this plot is in file \"MDSclustering_WholeBrain_RUVremoved.html\"\n",
      "\n",
      "\n",
      "```{r Figure 7, fig.cap = \"Figure 7. Multidimensional Scaling on the top 5,000 most variable genes after removing the effects of 4 RUV factors\", echo = F, message=FALSE, fig.width = 7, fig.height = 6}\n",
      "ggplot(plot.values4, aes(Dim1, Dim2, color = Trt))+\n",
      "  geom_point(aes(fill = Trt),  size = 4)+theme_bw()+\n",
      "  xlab(paste0(\"Dimension 1 \",\"(\",plot.values4$var.exp[1],\"%)\")) +\n",
      "  ylab(paste0(\"Dimension 2 \",\"(\",plot.values4$var.exp[2],\"%)\")) +\n",
      "#  geom_text(aes(label = Rep),hjust=0.5, vjust=-0.6) +  xlim(-0.5, 0.3) +\n",
      "  scale_color_manual(values = dittoSeq::dittoColors()[1:4])\n",
      "\n",
      "glimmaMDS(wobj_final$normalizedCounts, top = 5000, labels = d$samples$Label,\n",
      "          groups = d.filt$samples, html = \"MDSclustering_WholeBrain_RUVremoved.html\")\n",
      "\n",
      "\n",
      "\n",
      "```\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "## Differential gene expression analysis - RUV\n",
      "\n",
      "Differential gene expression (DE) analysis was performed using the limma-trend method[^trend1] using a model of ~ treatment + 4 RUV factors. The same pairwise comparison was pulled from the model and FDR correction done. Now the lowest FDR p-value for any gene was `r min(results.out$adj.P.Val)` and `r sum(results.out$adj.P.Val < 0.05)` genes had FDR p-value < 0.05. We can further increase the number of genes to examine by increasing the FDR threshold to 0.1, which yields `r sum(codes_fdr0.1_ruv[,1] == 1)` genes with higher expression in Flu vs. Veh and `r sum(codes_fdr0.1_ruv[,1] == -1)` genes with lower expression in Flu vs. Veh. These genes are visualized in a heatmap (Figure 8). \n",
      "\n",
      "\n",
      "```{r Figure 8, fig.cap = \"Figure 8. Heatmap of 2338 genes with onweay ANOVA FDR p-value < 0.1\", echo = F, message=FALSE, fig.width = 10, fig.height = 10, cache = TRUE}\n",
      "\n",
      "heatmap.2(heat.h2, col=col.pan, Rowv=as.dendrogram(x.cluster.h2),Colv=FALSE, scale=\"none\",\n",
      "          trace=\"none\", dendrogram=\"row\", cexRow=0.9, cexCol=1.5, labRow = \"\",\n",
      "          keysize = 0.2,lwid = c(1.4,4), lhei = c(1,4), key.par=list(mar = c(4,2,2,1)),\n",
      "          margins = c(9,2),density.info = \"none\", key.xlab = \"SD from mean\",\n",
      "          colsep =c(4), sepcolor=\"black\",\n",
      "          sepwidth=c(0.05,0.05))\n",
      "\n",
      "\n",
      "\n",
      "```\n",
      "\n",
      "\n",
      "### Interactive results - RUV\n",
      "\n",
      "Below is an interactive version of the Flu vs. Veh results using RUV factors. A stand-alone interactive .html file is also in the main output directory. There is also non stand-alone .html files of similar interactive results available in the \"interactive_plots\" directory. This is an older version that are better in a few ways (quicker switching to top right plot and auto-scaling its y-axis) but do not contain all of the functionality of the newer interactive plots. In the old version the **AdjPValue** column actually gives the raw p-values whereas both raw and FDR corrected p-values are in the new versions. \n",
      "\n",
      "\n",
      "\n",
      "```{r interactive2, echo = FALSE, eval = TRUE}\n",
      "\n",
      "i <- 1\n",
      "\n",
      "#annot <- data.frame(results.out0[,c(2:3, 7)])\n",
      "glimmaMA(fit.ruv4b, d.filt, counts = wobj_final$normalizedCounts, \n",
      "         transform.counts = \"none\",\n",
      "         groups =d.filt$samples$Trt, coef = i,\n",
      "         status  = codes_fdr0.1_ruv, \n",
      "         main = paste(colnames(codes_fdr0.1_ruv)[i], \"RUV\"),\n",
      ")\n",
      "         #anno = annot)\n",
      "\n",
      "glimmaMA(fit.ruv4b, d.filt, counts = wobj_final$normalizedCounts, \n",
      "         transform.counts = \"none\",\n",
      "         groups =d.filt$samples$Trt, coef = i,\n",
      "         status  = codes_fdr0.1_ruv, \n",
      "         main = paste(colnames(codes_fdr0.1_ruv)[i], \"RUV\"),\n",
      "         html = \"WholeBrain_RUV_FDR0.1.html\")\n",
      "\n",
      "\n",
      "```\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "## Hardware and software descriptions\n",
      "The read quality check and count generation was done using CNRG's [*Biocluster* high-performance computing resource](http://biocluster.igb.illinois.edu/). All analyses from summation of counts to the gene level (including the codes in .Rmd file) were done on a laptop in `r R.Version()$version.string`[^R] using packages as indicated below.\n",
      "\n",
      "\n",
      "## R and package versions\n",
      "\n",
      "```{r}\n",
      "sessionInfo()\n",
      "```\n",
      "\n",
      "\n",
      "\n",
      "## References\n",
      "\n",
      "[^R]: R Core Team (2017). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL https://www.R-project.org/.\n",
      "\n",
      "[^tmm]: Robinson MD, Oshlack A (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology 11, R25.\n",
      "\n",
      "[^limma]: Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.\n",
      "\n",
      "[^voom]: Law CW, Chen Y, Shi W, et al.: voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biol. 2014; 15(2): R29\n",
      "\n",
      "[^fdr]: Benjamini, Y., & Hochberg, Y. (1995). Controlling the false discovery rate: a practical and powerful approach to multiple testing. Journal of the Royal statistical society: series B (Methodological), 57(1), 289-300.\n",
      "\n",
      "[^voom]: Law, C. W., Chen, Y., Shi, W., & Smyth, G. K. (2014). voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome biology, 15(2), R29.\n",
      "\n",
      "[^treat]: McCarthy, D. J., & Smyth, G. K. (2009). Testing significance relative to a fold-change threshold is a TREAT. Bioinformatics, 25(6), 765-771.\n",
      "\n",
      "[^annohub]: Carlson M (2019). org.Gg.eg.db: Genome wide annotation for Mouse. R package version 3.8.2.\n",
      "\n",
      "[^bowtie2]: Langmead B, Salzberg S. Fast gapped-read alignment with Bowtie 2. Nature Methods. 2012, 9:357-359.\n",
      "\n",
      "[^trimmomatic]: Bolger, A. M., Lohse, M., & Usadel, B. (2014). Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics, 30(15), 2114-2120.\n",
      "\n",
      "[^trend1]: Chen Y, Lun ATL and Smyth GK. From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline [version 2; referees: 5 approved]. F1000Research 2016, 5:1438 (doi: 10.12688/f1000research.8987.2)\n",
      "\n",
      "[^tximport]: Soneson C, Love MI, Robinson MD (2015). \"Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences.\" F1000Research, 4. doi: 10.12688/f1000research.7563.1.\n",
      "\n",
      "[^salmon]: Patro, R., Duggal, G., Love, M. I., Irizarry, R. A., & Kingsford, C. (2017). Salmon provides fast and bias-aware quantification of transcript expression. Nature methods, 14(4), 417.\n",
      "\n",
      "[^fastqc]: Andrews, S. (2010). FastQC: a quality control tool for high throughput sequence data. Available at: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\n",
      "\n",
      "[^multiqc]: Ewels, P., Magnusson, M., Lundin, S., & K?ller, M. (2016). MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics, 32(19), 3047-3048.\n",
      "\n",
      "[^sva]: Leek JT, Johnson WE, Parker HS, Fertig EJ, Jaffe AE, Storey JD, Zhang Y, Torres LC (2019). sva: Surrogate Variable Analysis. R package version 3.32.0.\n",
      "\n",
      "[^bioc]: Huber W., Carey V.J., Gentleman R., ..., Morgan M. Orchestrating high-throughput genomic analysis with Bioconductor. Nature Methods, 2015:12, 115.\n",
      "\n",
      "[^edger1]: McCarthy DJ, Chen Y, Smyth GK (2012). \"Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation.\" Nucleic Acids Research, 40(10), 4288-4297.\n",
      "\n",
      "[^edger2]: Robinson MD, McCarthy DJ, Smyth GK (2010). \"edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.\" Bioinformatics, 26(1), 139-140.\n",
      "\n",
      "[^edger3]: Chen Y, Lun ATL and Smyth GK. From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline [version 2; peer review: 5 approved]. F1000Research 2016, 5:1438 (https://doi.org/10.12688/f1000research.8987.2)\n",
      "\n",
      "[^ruvseq]: Risso D, Ngai J, Speed T, Dudoit S (2014). \"Normalization of RNA-seq data using factor analysis of control genes or samples.\" Nature Biotechnology, 32(9), 896-902.\n",
      "\n",
      "[^star]: Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P, Chaisson M, Gingeras TR. STAR: ultrafast universal RNA-seq aligner. Bioinformatics. 2013 Jan 1;29(1):15-21. doi: 10.1093/bioinformatics/bts635. Epub 2012 Oct 25. PMID: 23104886; PMCID: PMC3530905.\n",
      "\n",
      "[^fc]: Liao Y, Smyth GK and Shi W. featureCounts: an efficient general-purpose program for assigning sequence reads to genomic features. Bioinformatics, 30(7):923-30, 2014.\n",
      "\n",
      "[^wgcna1]: Bin Zhang and Steve Horvath (2005) A General Framework for Weighted Gene Co-Expression Network Analysis. Statistical Applications in Genetics and Molecular Biology: Vol. 4: No. 1, Article 17 PMID: 16646834\n",
      "\n",
      "[^wgcna2]: Horvath, S. (2008). WGCNA: an R package for weighted correlation network analysis. BMC Bioinformatics., 9(1), 559. https://doi.org/10.1186/1471-2105-9-559\n",
      "\n",
      "[^iterRUV]: Peixoto, L., Risso, D., Poplawski, S. G., Wimmer, M. E., Speed, T. P., Wood, M. A., & Abel, T. (2015). How data analysis affects power, reproducibility and biological insight of RNA-seq studies in complex datasets. Nucleic acids research, 43(16), 7664–7674. https://doi.org/10.1093/nar/gkv736\n",
      "<END OF FILE>\u001b[0m\n",
      "Thought:"
     ]
    },
    {
     "ename": "InvalidRequestError",
     "evalue": "This model's maximum context length is 8192 tokens. However, your messages resulted in 8199 tokens. Please reduce the length of the messages.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInvalidRequestError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m/Users/kastanday/code/ncsa/ai-ta/langchain-improved-agents/docs/docs_skeleton/docs/integrations/toolkits/github.ipynb Cell 17\u001b[0m line \u001b[0;36m3\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/kastanday/code/ncsa/ai-ta/langchain-improved-agents/docs/docs_skeleton/docs/integrations/toolkits/github.ipynb#Y135sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mlangchain\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/kastanday/code/ncsa/ai-ta/langchain-improved-agents/docs/docs_skeleton/docs/integrations/toolkits/github.ipynb#Y135sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m langchain\u001b[39m.\u001b[39mdebug \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/kastanday/code/ncsa/ai-ta/langchain-improved-agents/docs/docs_skeleton/docs/integrations/toolkits/github.ipynb#Y135sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m agent\u001b[39m.\u001b[39;49mrun(\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/kastanday/code/ncsa/ai-ta/langchain-improved-agents/docs/docs_skeleton/docs/integrations/toolkits/github.ipynb#Y135sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39mYou have the software engineering capabilities of a Google Principle engineer. You are tasked with searching the code in this repo for files related to RNA-seq, particularly fastq code. Please reply with the filenames and relevant code snippets.\u001b[39;49m\u001b[39m\"\u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/kastanday/code/ncsa/ai-ta/langchain-improved-agents/docs/docs_skeleton/docs/integrations/toolkits/github.ipynb#Y135sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m )\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/base.py:501\u001b[0m, in \u001b[0;36mChain.run\u001b[0;34m(self, callbacks, tags, metadata, *args, **kwargs)\u001b[0m\n\u001b[1;32m    499\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(args) \u001b[39m!=\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m    500\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39m`run` supports only one positional argument.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 501\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m(args[\u001b[39m0\u001b[39;49m], callbacks\u001b[39m=\u001b[39;49mcallbacks, tags\u001b[39m=\u001b[39;49mtags, metadata\u001b[39m=\u001b[39;49mmetadata)[\n\u001b[1;32m    502\u001b[0m         _output_key\n\u001b[1;32m    503\u001b[0m     ]\n\u001b[1;32m    505\u001b[0m \u001b[39mif\u001b[39;00m kwargs \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m args:\n\u001b[1;32m    506\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m(kwargs, callbacks\u001b[39m=\u001b[39mcallbacks, tags\u001b[39m=\u001b[39mtags, metadata\u001b[39m=\u001b[39mmetadata)[\n\u001b[1;32m    507\u001b[0m         _output_key\n\u001b[1;32m    508\u001b[0m     ]\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/base.py:306\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    304\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    305\u001b[0m     run_manager\u001b[39m.\u001b[39mon_chain_error(e)\n\u001b[0;32m--> 306\u001b[0m     \u001b[39mraise\u001b[39;00m e\n\u001b[1;32m    307\u001b[0m run_manager\u001b[39m.\u001b[39mon_chain_end(outputs)\n\u001b[1;32m    308\u001b[0m final_outputs: Dict[\u001b[39mstr\u001b[39m, Any] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprep_outputs(\n\u001b[1;32m    309\u001b[0m     inputs, outputs, return_only_outputs\n\u001b[1;32m    310\u001b[0m )\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/base.py:300\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    293\u001b[0m run_manager \u001b[39m=\u001b[39m callback_manager\u001b[39m.\u001b[39mon_chain_start(\n\u001b[1;32m    294\u001b[0m     dumpd(\u001b[39mself\u001b[39m),\n\u001b[1;32m    295\u001b[0m     inputs,\n\u001b[1;32m    296\u001b[0m     name\u001b[39m=\u001b[39mrun_name,\n\u001b[1;32m    297\u001b[0m )\n\u001b[1;32m    298\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    299\u001b[0m     outputs \u001b[39m=\u001b[39m (\n\u001b[0;32m--> 300\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_call(inputs, run_manager\u001b[39m=\u001b[39;49mrun_manager)\n\u001b[1;32m    301\u001b[0m         \u001b[39mif\u001b[39;00m new_arg_supported\n\u001b[1;32m    302\u001b[0m         \u001b[39melse\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_call(inputs)\n\u001b[1;32m    303\u001b[0m     )\n\u001b[1;32m    304\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    305\u001b[0m     run_manager\u001b[39m.\u001b[39mon_chain_error(e)\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/agents/agent.py:1141\u001b[0m, in \u001b[0;36mAgentExecutor._call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m   1139\u001b[0m \u001b[39m# We now enter the agent loop (until it returns something).\u001b[39;00m\n\u001b[1;32m   1140\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_should_continue(iterations, time_elapsed):\n\u001b[0;32m-> 1141\u001b[0m     next_step_output \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_take_next_step(\n\u001b[1;32m   1142\u001b[0m         name_to_tool_map,\n\u001b[1;32m   1143\u001b[0m         color_mapping,\n\u001b[1;32m   1144\u001b[0m         inputs,\n\u001b[1;32m   1145\u001b[0m         intermediate_steps,\n\u001b[1;32m   1146\u001b[0m         run_manager\u001b[39m=\u001b[39;49mrun_manager,\n\u001b[1;32m   1147\u001b[0m     )\n\u001b[1;32m   1148\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(next_step_output, AgentFinish):\n\u001b[1;32m   1149\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_return(\n\u001b[1;32m   1150\u001b[0m             next_step_output, intermediate_steps, run_manager\u001b[39m=\u001b[39mrun_manager\n\u001b[1;32m   1151\u001b[0m         )\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/agents/agent.py:928\u001b[0m, in \u001b[0;36mAgentExecutor._take_next_step\u001b[0;34m(self, name_to_tool_map, color_mapping, inputs, intermediate_steps, run_manager)\u001b[0m\n\u001b[1;32m    925\u001b[0m     intermediate_steps \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_prepare_intermediate_steps(intermediate_steps)\n\u001b[1;32m    927\u001b[0m     \u001b[39m# Call the LLM to see what to do.\u001b[39;00m\n\u001b[0;32m--> 928\u001b[0m     output \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49magent\u001b[39m.\u001b[39;49mplan(\n\u001b[1;32m    929\u001b[0m         intermediate_steps,\n\u001b[1;32m    930\u001b[0m         callbacks\u001b[39m=\u001b[39;49mrun_manager\u001b[39m.\u001b[39;49mget_child() \u001b[39mif\u001b[39;49;00m run_manager \u001b[39melse\u001b[39;49;00m \u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m    931\u001b[0m         \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49minputs,\n\u001b[1;32m    932\u001b[0m     )\n\u001b[1;32m    933\u001b[0m \u001b[39mexcept\u001b[39;00m OutputParserException \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    934\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandle_parsing_errors, \u001b[39mbool\u001b[39m):\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/agents/agent.py:541\u001b[0m, in \u001b[0;36mAgent.plan\u001b[0;34m(self, intermediate_steps, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    529\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Given input, decided what to do.\u001b[39;00m\n\u001b[1;32m    530\u001b[0m \n\u001b[1;32m    531\u001b[0m \u001b[39mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    538\u001b[0m \u001b[39m    Action specifying what tool to use.\u001b[39;00m\n\u001b[1;32m    539\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    540\u001b[0m full_inputs \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_full_inputs(intermediate_steps, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m--> 541\u001b[0m full_output \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mllm_chain\u001b[39m.\u001b[39;49mpredict(callbacks\u001b[39m=\u001b[39;49mcallbacks, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mfull_inputs)\n\u001b[1;32m    542\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39moutput_parser\u001b[39m.\u001b[39mparse(full_output)\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/llm.py:257\u001b[0m, in \u001b[0;36mLLMChain.predict\u001b[0;34m(self, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpredict\u001b[39m(\u001b[39mself\u001b[39m, callbacks: Callbacks \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs: Any) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mstr\u001b[39m:\n\u001b[1;32m    243\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Format prompt with kwargs and pass to LLM.\u001b[39;00m\n\u001b[1;32m    244\u001b[0m \n\u001b[1;32m    245\u001b[0m \u001b[39m    Args:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    255\u001b[0m \u001b[39m            completion = llm.predict(adjective=\"funny\")\u001b[39;00m\n\u001b[1;32m    256\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 257\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m(kwargs, callbacks\u001b[39m=\u001b[39;49mcallbacks)[\u001b[39mself\u001b[39m\u001b[39m.\u001b[39moutput_key]\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/base.py:306\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    304\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    305\u001b[0m     run_manager\u001b[39m.\u001b[39mon_chain_error(e)\n\u001b[0;32m--> 306\u001b[0m     \u001b[39mraise\u001b[39;00m e\n\u001b[1;32m    307\u001b[0m run_manager\u001b[39m.\u001b[39mon_chain_end(outputs)\n\u001b[1;32m    308\u001b[0m final_outputs: Dict[\u001b[39mstr\u001b[39m, Any] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprep_outputs(\n\u001b[1;32m    309\u001b[0m     inputs, outputs, return_only_outputs\n\u001b[1;32m    310\u001b[0m )\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/base.py:300\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    293\u001b[0m run_manager \u001b[39m=\u001b[39m callback_manager\u001b[39m.\u001b[39mon_chain_start(\n\u001b[1;32m    294\u001b[0m     dumpd(\u001b[39mself\u001b[39m),\n\u001b[1;32m    295\u001b[0m     inputs,\n\u001b[1;32m    296\u001b[0m     name\u001b[39m=\u001b[39mrun_name,\n\u001b[1;32m    297\u001b[0m )\n\u001b[1;32m    298\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    299\u001b[0m     outputs \u001b[39m=\u001b[39m (\n\u001b[0;32m--> 300\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_call(inputs, run_manager\u001b[39m=\u001b[39;49mrun_manager)\n\u001b[1;32m    301\u001b[0m         \u001b[39mif\u001b[39;00m new_arg_supported\n\u001b[1;32m    302\u001b[0m         \u001b[39melse\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_call(inputs)\n\u001b[1;32m    303\u001b[0m     )\n\u001b[1;32m    304\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    305\u001b[0m     run_manager\u001b[39m.\u001b[39mon_chain_error(e)\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/llm.py:93\u001b[0m, in \u001b[0;36mLLMChain._call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m     88\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_call\u001b[39m(\n\u001b[1;32m     89\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m     90\u001b[0m     inputs: Dict[\u001b[39mstr\u001b[39m, Any],\n\u001b[1;32m     91\u001b[0m     run_manager: Optional[CallbackManagerForChainRun] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m     92\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Dict[\u001b[39mstr\u001b[39m, \u001b[39mstr\u001b[39m]:\n\u001b[0;32m---> 93\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgenerate([inputs], run_manager\u001b[39m=\u001b[39;49mrun_manager)\n\u001b[1;32m     94\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcreate_outputs(response)[\u001b[39m0\u001b[39m]\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chains/llm.py:103\u001b[0m, in \u001b[0;36mLLMChain.generate\u001b[0;34m(self, input_list, run_manager)\u001b[0m\n\u001b[1;32m    101\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Generate LLM result from inputs.\"\"\"\u001b[39;00m\n\u001b[1;32m    102\u001b[0m prompts, stop \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprep_prompts(input_list, run_manager\u001b[39m=\u001b[39mrun_manager)\n\u001b[0;32m--> 103\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mllm\u001b[39m.\u001b[39;49mgenerate_prompt(\n\u001b[1;32m    104\u001b[0m     prompts,\n\u001b[1;32m    105\u001b[0m     stop,\n\u001b[1;32m    106\u001b[0m     callbacks\u001b[39m=\u001b[39;49mrun_manager\u001b[39m.\u001b[39;49mget_child() \u001b[39mif\u001b[39;49;00m run_manager \u001b[39melse\u001b[39;49;00m \u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m    107\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mllm_kwargs,\n\u001b[1;32m    108\u001b[0m )\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/base.py:469\u001b[0m, in \u001b[0;36mBaseChatModel.generate_prompt\u001b[0;34m(self, prompts, stop, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    461\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mgenerate_prompt\u001b[39m(\n\u001b[1;32m    462\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m    463\u001b[0m     prompts: List[PromptValue],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    466\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs: Any,\n\u001b[1;32m    467\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m LLMResult:\n\u001b[1;32m    468\u001b[0m     prompt_messages \u001b[39m=\u001b[39m [p\u001b[39m.\u001b[39mto_messages() \u001b[39mfor\u001b[39;00m p \u001b[39min\u001b[39;00m prompts]\n\u001b[0;32m--> 469\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgenerate(prompt_messages, stop\u001b[39m=\u001b[39;49mstop, callbacks\u001b[39m=\u001b[39;49mcallbacks, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/base.py:359\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, **kwargs)\u001b[0m\n\u001b[1;32m    357\u001b[0m         \u001b[39mif\u001b[39;00m run_managers:\n\u001b[1;32m    358\u001b[0m             run_managers[i]\u001b[39m.\u001b[39mon_llm_error(e)\n\u001b[0;32m--> 359\u001b[0m         \u001b[39mraise\u001b[39;00m e\n\u001b[1;32m    360\u001b[0m flattened_outputs \u001b[39m=\u001b[39m [\n\u001b[1;32m    361\u001b[0m     LLMResult(generations\u001b[39m=\u001b[39m[res\u001b[39m.\u001b[39mgenerations], llm_output\u001b[39m=\u001b[39mres\u001b[39m.\u001b[39mllm_output)\n\u001b[1;32m    362\u001b[0m     \u001b[39mfor\u001b[39;00m res \u001b[39min\u001b[39;00m results\n\u001b[1;32m    363\u001b[0m ]\n\u001b[1;32m    364\u001b[0m llm_output \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_combine_llm_outputs([res\u001b[39m.\u001b[39mllm_output \u001b[39mfor\u001b[39;00m res \u001b[39min\u001b[39;00m results])\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/base.py:349\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, **kwargs)\u001b[0m\n\u001b[1;32m    346\u001b[0m \u001b[39mfor\u001b[39;00m i, m \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(messages):\n\u001b[1;32m    347\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    348\u001b[0m         results\u001b[39m.\u001b[39mappend(\n\u001b[0;32m--> 349\u001b[0m             \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_generate_with_cache(\n\u001b[1;32m    350\u001b[0m                 m,\n\u001b[1;32m    351\u001b[0m                 stop\u001b[39m=\u001b[39;49mstop,\n\u001b[1;32m    352\u001b[0m                 run_manager\u001b[39m=\u001b[39;49mrun_managers[i] \u001b[39mif\u001b[39;49;00m run_managers \u001b[39melse\u001b[39;49;00m \u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m    353\u001b[0m                 \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m    354\u001b[0m             )\n\u001b[1;32m    355\u001b[0m         )\n\u001b[1;32m    356\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    357\u001b[0m         \u001b[39mif\u001b[39;00m run_managers:\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/base.py:501\u001b[0m, in \u001b[0;36mBaseChatModel._generate_with_cache\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    497\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    498\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mAsked to cache, but no cache found at `langchain.cache`.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    499\u001b[0m     )\n\u001b[1;32m    500\u001b[0m \u001b[39mif\u001b[39;00m new_arg_supported:\n\u001b[0;32m--> 501\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_generate(\n\u001b[1;32m    502\u001b[0m         messages, stop\u001b[39m=\u001b[39;49mstop, run_manager\u001b[39m=\u001b[39;49mrun_manager, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs\n\u001b[1;32m    503\u001b[0m     )\n\u001b[1;32m    504\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    505\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_generate(messages, stop\u001b[39m=\u001b[39mstop, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/openai.py:403\u001b[0m, in \u001b[0;36mChatOpenAI._generate\u001b[0;34m(self, messages, stop, run_manager, stream, **kwargs)\u001b[0m\n\u001b[1;32m    401\u001b[0m message_dicts, params \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_create_message_dicts(messages, stop)\n\u001b[1;32m    402\u001b[0m params \u001b[39m=\u001b[39m {\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mparams, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs}\n\u001b[0;32m--> 403\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcompletion_with_retry(\n\u001b[1;32m    404\u001b[0m     messages\u001b[39m=\u001b[39;49mmessage_dicts, run_manager\u001b[39m=\u001b[39;49mrun_manager, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mparams\n\u001b[1;32m    405\u001b[0m )\n\u001b[1;32m    406\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_create_chat_result(response)\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/openai.py:282\u001b[0m, in \u001b[0;36mChatOpenAI.completion_with_retry\u001b[0;34m(self, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39m@retry_decorator\u001b[39m\n\u001b[1;32m    279\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_completion_with_retry\u001b[39m(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs: Any) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Any:\n\u001b[1;32m    280\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclient\u001b[39m.\u001b[39mcreate(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m--> 282\u001b[0m \u001b[39mreturn\u001b[39;00m _completion_with_retry(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/tenacity/__init__.py:289\u001b[0m, in \u001b[0;36mBaseRetrying.wraps.<locals>.wrapped_f\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    287\u001b[0m \u001b[39m@functools\u001b[39m\u001b[39m.\u001b[39mwraps(f)\n\u001b[1;32m    288\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwrapped_f\u001b[39m(\u001b[39m*\u001b[39margs: t\u001b[39m.\u001b[39mAny, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkw: t\u001b[39m.\u001b[39mAny) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m t\u001b[39m.\u001b[39mAny:\n\u001b[0;32m--> 289\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m(f, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkw)\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/tenacity/__init__.py:379\u001b[0m, in \u001b[0;36mRetrying.__call__\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    377\u001b[0m retry_state \u001b[39m=\u001b[39m RetryCallState(retry_object\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m, fn\u001b[39m=\u001b[39mfn, args\u001b[39m=\u001b[39margs, kwargs\u001b[39m=\u001b[39mkwargs)\n\u001b[1;32m    378\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 379\u001b[0m     do \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49miter(retry_state\u001b[39m=\u001b[39;49mretry_state)\n\u001b[1;32m    380\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(do, DoAttempt):\n\u001b[1;32m    381\u001b[0m         \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/tenacity/__init__.py:314\u001b[0m, in \u001b[0;36mBaseRetrying.iter\u001b[0;34m(self, retry_state)\u001b[0m\n\u001b[1;32m    312\u001b[0m is_explicit_retry \u001b[39m=\u001b[39m fut\u001b[39m.\u001b[39mfailed \u001b[39mand\u001b[39;00m \u001b[39misinstance\u001b[39m(fut\u001b[39m.\u001b[39mexception(), TryAgain)\n\u001b[1;32m    313\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m (is_explicit_retry \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mretry(retry_state)):\n\u001b[0;32m--> 314\u001b[0m     \u001b[39mreturn\u001b[39;00m fut\u001b[39m.\u001b[39;49mresult()\n\u001b[1;32m    316\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mafter \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    317\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mafter(retry_state)\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/concurrent/futures/_base.py:451\u001b[0m, in \u001b[0;36mFuture.result\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    449\u001b[0m     \u001b[39mraise\u001b[39;00m CancelledError()\n\u001b[1;32m    450\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m==\u001b[39m FINISHED:\n\u001b[0;32m--> 451\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__get_result()\n\u001b[1;32m    453\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_condition\u001b[39m.\u001b[39mwait(timeout)\n\u001b[1;32m    455\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39min\u001b[39;00m [CANCELLED, CANCELLED_AND_NOTIFIED]:\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/concurrent/futures/_base.py:403\u001b[0m, in \u001b[0;36mFuture.__get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    401\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exception:\n\u001b[1;32m    402\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 403\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exception\n\u001b[1;32m    404\u001b[0m     \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m    405\u001b[0m         \u001b[39m# Break a reference cycle with the exception in self._exception\u001b[39;00m\n\u001b[1;32m    406\u001b[0m         \u001b[39mself\u001b[39m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/tenacity/__init__.py:382\u001b[0m, in \u001b[0;36mRetrying.__call__\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    380\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(do, DoAttempt):\n\u001b[1;32m    381\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 382\u001b[0m         result \u001b[39m=\u001b[39m fn(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    383\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m:  \u001b[39m# noqa: B902\u001b[39;00m\n\u001b[1;32m    384\u001b[0m         retry_state\u001b[39m.\u001b[39mset_exception(sys\u001b[39m.\u001b[39mexc_info())  \u001b[39m# type: ignore[arg-type]\u001b[39;00m\n",
      "File \u001b[0;32m~/code/ncsa/ai-ta/langchain-improved-agents/libs/langchain/langchain/chat_models/openai.py:280\u001b[0m, in \u001b[0;36mChatOpenAI.completion_with_retry.<locals>._completion_with_retry\u001b[0;34m(**kwargs)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39m@retry_decorator\u001b[39m\n\u001b[1;32m    279\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_completion_with_retry\u001b[39m(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs: Any) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Any:\n\u001b[0;32m--> 280\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mclient\u001b[39m.\u001b[39;49mcreate(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/openai/api_resources/chat_completion.py:25\u001b[0m, in \u001b[0;36mChatCompletion.create\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m     24\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 25\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49mcreate(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     26\u001b[0m     \u001b[39mexcept\u001b[39;00m TryAgain \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m     27\u001b[0m         \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m time\u001b[39m.\u001b[39mtime() \u001b[39m>\u001b[39m start \u001b[39m+\u001b[39m timeout:\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py:155\u001b[0m, in \u001b[0;36mEngineAPIResource.create\u001b[0;34m(cls, api_key, api_base, api_type, request_id, api_version, organization, **params)\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[1;32m    130\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcreate\u001b[39m(\n\u001b[1;32m    131\u001b[0m     \u001b[39mcls\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    138\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mparams,\n\u001b[1;32m    139\u001b[0m ):\n\u001b[1;32m    140\u001b[0m     (\n\u001b[1;32m    141\u001b[0m         deployment_id,\n\u001b[1;32m    142\u001b[0m         engine,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    152\u001b[0m         api_key, api_base, api_type, api_version, organization, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mparams\n\u001b[1;32m    153\u001b[0m     )\n\u001b[0;32m--> 155\u001b[0m     response, _, api_key \u001b[39m=\u001b[39m requestor\u001b[39m.\u001b[39;49mrequest(\n\u001b[1;32m    156\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    157\u001b[0m         url,\n\u001b[1;32m    158\u001b[0m         params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    159\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    160\u001b[0m         stream\u001b[39m=\u001b[39;49mstream,\n\u001b[1;32m    161\u001b[0m         request_id\u001b[39m=\u001b[39;49mrequest_id,\n\u001b[1;32m    162\u001b[0m         request_timeout\u001b[39m=\u001b[39;49mrequest_timeout,\n\u001b[1;32m    163\u001b[0m     )\n\u001b[1;32m    165\u001b[0m     \u001b[39mif\u001b[39;00m stream:\n\u001b[1;32m    166\u001b[0m         \u001b[39m# must be an iterator\u001b[39;00m\n\u001b[1;32m    167\u001b[0m         \u001b[39massert\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(response, OpenAIResponse)\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/openai/api_requestor.py:299\u001b[0m, in \u001b[0;36mAPIRequestor.request\u001b[0;34m(self, method, url, params, headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mrequest\u001b[39m(\n\u001b[1;32m    279\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m    280\u001b[0m     method,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    287\u001b[0m     request_timeout: Optional[Union[\u001b[39mfloat\u001b[39m, Tuple[\u001b[39mfloat\u001b[39m, \u001b[39mfloat\u001b[39m]]] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    288\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Tuple[Union[OpenAIResponse, Iterator[OpenAIResponse]], \u001b[39mbool\u001b[39m, \u001b[39mstr\u001b[39m]:\n\u001b[1;32m    289\u001b[0m     result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mrequest_raw(\n\u001b[1;32m    290\u001b[0m         method\u001b[39m.\u001b[39mlower(),\n\u001b[1;32m    291\u001b[0m         url,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    297\u001b[0m         request_timeout\u001b[39m=\u001b[39mrequest_timeout,\n\u001b[1;32m    298\u001b[0m     )\n\u001b[0;32m--> 299\u001b[0m     resp, got_stream \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_interpret_response(result, stream)\n\u001b[1;32m    300\u001b[0m     \u001b[39mreturn\u001b[39;00m resp, got_stream, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapi_key\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/openai/api_requestor.py:710\u001b[0m, in \u001b[0;36mAPIRequestor._interpret_response\u001b[0;34m(self, result, stream)\u001b[0m\n\u001b[1;32m    702\u001b[0m     \u001b[39mreturn\u001b[39;00m (\n\u001b[1;32m    703\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_interpret_response_line(\n\u001b[1;32m    704\u001b[0m             line, result\u001b[39m.\u001b[39mstatus_code, result\u001b[39m.\u001b[39mheaders, stream\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    705\u001b[0m         )\n\u001b[1;32m    706\u001b[0m         \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m parse_stream(result\u001b[39m.\u001b[39miter_lines())\n\u001b[1;32m    707\u001b[0m     ), \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    708\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    709\u001b[0m     \u001b[39mreturn\u001b[39;00m (\n\u001b[0;32m--> 710\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_interpret_response_line(\n\u001b[1;32m    711\u001b[0m             result\u001b[39m.\u001b[39;49mcontent\u001b[39m.\u001b[39;49mdecode(\u001b[39m\"\u001b[39;49m\u001b[39mutf-8\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m    712\u001b[0m             result\u001b[39m.\u001b[39;49mstatus_code,\n\u001b[1;32m    713\u001b[0m             result\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    714\u001b[0m             stream\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    715\u001b[0m         ),\n\u001b[1;32m    716\u001b[0m         \u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    717\u001b[0m     )\n",
      "File \u001b[0;32m~/miniforge3/envs/flask10_py10/lib/python3.10/site-packages/openai/api_requestor.py:775\u001b[0m, in \u001b[0;36mAPIRequestor._interpret_response_line\u001b[0;34m(self, rbody, rcode, rheaders, stream)\u001b[0m\n\u001b[1;32m    773\u001b[0m stream_error \u001b[39m=\u001b[39m stream \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39merror\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m resp\u001b[39m.\u001b[39mdata\n\u001b[1;32m    774\u001b[0m \u001b[39mif\u001b[39;00m stream_error \u001b[39mor\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39m200\u001b[39m \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m rcode \u001b[39m<\u001b[39m \u001b[39m300\u001b[39m:\n\u001b[0;32m--> 775\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandle_error_response(\n\u001b[1;32m    776\u001b[0m         rbody, rcode, resp\u001b[39m.\u001b[39mdata, rheaders, stream_error\u001b[39m=\u001b[39mstream_error\n\u001b[1;32m    777\u001b[0m     )\n\u001b[1;32m    778\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "\u001b[0;31mInvalidRequestError\u001b[0m: This model's maximum context length is 8192 tokens. However, your messages resulted in 8199 tokens. Please reduce the length of the messages."
     ]
    }
   ],
   "source": [
    "import langchain\n",
    "langchain.debug = False\n",
    "agent.run(\n",
    "    \"You have the software engineering capabilities of a Google Principle engineer. You are tasked with searching the code in this repo for files related to RNA-seq, particularly fastq code. Please reply with the filenames and relevant code snippets.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agent.run(\n",
    "    \"You have the software engineering capabilities of a Google Principle engineer. You are tasked with completing issues on a github repository. Please look at the existing issues and complete them. Make sure to create a new branch before making changes, then open a PR and create a review request for the user who opened the issue.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example: Advanced Agent\n",
    "\n",
    "If your agent does not need to use all 8 tools, you can build tools individually to use. For this example, we'll make an agent that does not use the create_file, delete_file or create_pull_request tools, but can also use duckduckgo-search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%pip install duckduckgo-search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.tools.github.tool import GitHubAction\n",
    "from langchain.tools import DuckDuckGoSearchRun\n",
    "from langchain.agents import Tool\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "\n",
    "tools = []\n",
    "unwanted_tools = ['Get Issue','Delete File', 'Create File', 'Create Pull Request']\n",
    "\n",
    "for tool in toolkit.get_tools():\n",
    "    if tool.name not in unwanted_tools:\n",
    "        tools.append(tool)\n",
    "tools+= [\n",
    "    Tool(\n",
    "        name = \"Search\",\n",
    "        func = DuckDuckGoSearchRun().run,\n",
    "        description = \"useful for when you need to search the web\"\n",
    "    )]\n",
    "        \n",
    "agent = initialize_agent(\n",
    "    tools = tools,\n",
    "    llm = ChatOpenAI(temperature=0.1),\n",
    "    agent = AgentType.ZERO_SHOT_REACT_DESCRIPTION,\n",
    "    verbose = True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally let's build a prompt and test it out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The GitHubAPIWrapper can be used outside of an agent, too\n",
    "# This gets the info about issue number 9, since we want to\n",
    "# force the agent to address this specific issue.\n",
    "\n",
    "issue = github.get_issue(9)\n",
    "\n",
    "prompt = f\"\"\"\n",
    "You are a senior frontend developer who is experienced in HTML, CSS, and JS especially React.\n",
    "You have been assigned the below issue. Complete it to the best of your ability.\n",
    "Remember to first make a plan and pay attention to details like file names and commonsense.\n",
    "Then execute the plan and use tools appropriately.\n",
    "Finally, make a pull request to merge your changes.\n",
    "Issue: {issue[\"title\"]}\n",
    "Issue Description: {issue['body']}\n",
    "Comments: {issue['comments']}\"\"\"\n",
    "\n",
    "agent.run(prompt)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
